# Scala Let's go!

In [1]:
import scala.io.Source

def loadPrimes(path: String = "primes.\$\$\$\$\$\$3\$\$\$\$\$\$") = {
    val lines = Source.fromFile(path).getLines
    lines.map(x => x.trim().toInt).toList
}

val primesOrdered = loadPrimes()
val primes = primesOrdered.toSet
val maxPrime = primes.max

def isPrime(n: Int): Boolean = n match {
    case x if x <= maxPrime => primes.contains(x)
    case _ => ???
}

import scala.io.Source


defined function loadPrimes
primesOrdered: List[Int] = List(
  2,
  3,
  5,
  7,
  11,
  13,
  17,
  19,
  23,
  29,
  31,
  37,
  41,
  43,
  47,
  53,
  59,
  61,
  67,
  71,
  73,
  79,
  83,
  89,
  97,
  101,
  103,
  107,
  109,
  113,
  127,
  131,
  137,
  139,
  149,
  151,
  157,
  163,
...
primes: Set[Int] = HashSet(
  744019,
  1237279,
  20873,
  276019,
  532949,
  620759,
  43261,
  1278373,
  999043,
  868559,
  1042949,
  504671,
  1161997,
  474223,
  1284037,
  37159,
  1241893,
  343939,
  14221,
  774791,
  553517,
  557891,
  850973,
  358219,
  587441,
  1173743,
  1055611,
  434179,
  1151729,
  1269197,
  39241,
  310889,
  69439,
  178889,
  789407,
  458407,
  219031,
  986023,
...
maxPrime: Int = 1299709
defined function isPrime

In [2]:
def collatz(N: Int): Int = N match {
    case 1 => 1
    case n if n % 2 == 0 => collatz(n / 2)
    case n => collatz(3 * n + 1)
}

def collatzOneStep(N: Int): Int = N match {
    case n if n % 2 == 0 => n / 2
    case n => 3 * n + 1
}

def getBase(N: Int): Int = N match {
    case n if n % 2 == 0 => getBase(n / 2)
    case n => n
}

object Path {
    def basePath(N: Int): List[Int] = N match {
        case 1 => List(1)
        case n if n % 2 == 0 => basePath(n / 2)
        case n => n :: basePath(3 * n + 1)
    }
    
    def primeLocked(N: Int): (List[Int], Boolean) = {
        val path = basePath(N)
        var result = List[Int]()

        var locked = isPrime(path.head)
        var success = true
        
        for (e <- path) {
            if (isPrime(e) == locked) {
                result = e :: result
            } else {
                result = e :: result
                
                if (locked && e != 1) {
                    success = false
                    return (result.reverse, success)
                }
                
                locked = isPrime(e)
            }
        }
        
        (result.reverse, success)
    }
    
    def findPrimeUnlock(N: Int) = {
        val path = basePath(N)
        
        // sliding window
        val result = path.sliding(2).toList
            .filter(e => isPrime(e.head) && e.last != 1 && !isPrime(e.last))
        
        if (result.isEmpty) {
            None
        } else {
            Some(result.head.last)
        }
    }
    
    def immediatePrimeUnlock(N: Int) = {
        val next = getBase(getBase(N) * 3 + 1)
        if (isPrime(N) && !isPrime(next) && next != 1) {
            Some(next)
        } else {
            None
        }
    }
}

defined function collatz
defined function collatzOneStep
defined function getBase
defined object Path

In [3]:
println(Path.basePath(71))
println(Path.immediatePrimeUnlock(107))

List(71, 107, 161, 121, 91, 137, 103, 155, 233, 175, 263, 395, 593, 445, 167, 251, 377, 283, 425, 319, 479, 719, 1079, 1619, 2429, 911, 1367, 2051, 3077, 577, 433, 325, 61, 23, 35, 53, 5, 1)
Some(161)


In [4]:
primesOrdered.foreach(prime => {
    val (path, success) = Path.primeLocked(prime)
    
    if (!success && path.length > 8) {
        println("Failed prime locking for " + prime)
        println(path)
    }
})

Failed prime locking for 5171
List(5171, 7757, 2909, 1091, 1637, 307, 461, 173, 65)
Failed prime locking for 13789
List(13789, 5171, 7757, 2909, 1091, 1637, 307, 461, 173, 65)


: 

## Finding the points at which the bases "unlock" from prime values

In [9]:
val problemPrimes = primesOrdered.take(1000)
    .filter(x => Path.immediatePrimeUnlock(x) != None)
    .map(x => (x, Path.immediatePrimeUnlock(x).get))

problemPrimes
    .foreach(x => println(s"Prime ${x._1} -> Composite ${x._2}"))

val sequence = problemPrimes.map(x => x._1).distinct.sorted
println(s"Problem primes: ${sequence}")


Prime 23 -> Composite 35
Prime 43 -> Composite 65
Prime 73 -> Composite 55
Prime 79 -> Composite 119
Prime 83 -> Composite 125
Prime 103 -> Composite 155
Prime 107 -> Composite 161
Prime 113 -> Composite 85
Prime 139 -> Composite 209
Prime 163 -> Composite 245
Prime 173 -> Composite 65
Prime 191 -> Composite 287
Prime 193 -> Composite 145
Prime 199 -> Composite 299
Prime 223 -> Composite 335
Prime 227 -> Composite 341
Prime 233 -> Composite 175
Prime 251 -> Composite 377
Prime 263 -> Composite 395
Prime 271 -> Composite 407
Prime 283 -> Composite 425
Prime 293 -> Composite 55
Prime 313 -> Composite 235
Prime 317 -> Composite 119
Prime 331 -> Composite 497
Prime 337 -> Composite 253
Prime 353 -> Composite 265
Prime 359 -> Composite 539
Prime 367 -> Composite 551
Prime 373 -> Composite 35
Prime 383 -> Composite 575
Prime 401 -> Composite 301
Prime 419 -> Composite 629
Prime 433 -> Composite 325
Prime 443 -> Composite 665
Prime 457 -> Composite 343
Prime 463 -> Composite 695
Prime 487 -> 

problemPrimes: List[(Int, Int)] = List(
  (23, 35),
  (43, 65),
  (73, 55),
  (79, 119),
  (83, 125),
  (103, 155),
  (107, 161),
  (113, 85),
  (139, 209),
  (163, 245),
  (173, 65),
  (191, 287),
  (193, 145),
  (199, 299),
  (223, 335),
  (227, 341),
  (233, 175),
  (251, 377),
  (263, 395),
  (271, 407),
  (283, 425),
  (293, 55),
  (313, 235),
  (317, 119),
  (331, 497),
  (337, 253),
  (353, 265),
  (359, 539),
  (367, 551),
  (373, 35),
  (383, 575),
  (401, 301),
  (419, 629),
  (433, 325),
  (443, 665),
  (457, 343),
  (463, 695),
  (487, 731),
...
sequence: List[Int] = List(
  23,
  43,
  73,
  79,
  83,
  103,
  107,
  113,
  139,
  163,
  173,
  191,
  193,
  199,
  223,
  227,
  233,
  251,
  263,
  271,
  283,
  293,
  313,
  317,
  331,
  337,
  353,
  359,
  367,
  373,
  383,
  401,
  419,
  433,
  443,
  457,
  463,
  487,
...

A "base" in the context of collatz is simply an odd number. Since any odd number can be doubled infinitely, the doubled numbers form an infinite
series of values that all collapse back to the odd number. In this way, "trunk bases" exist for sequences of numbers that necessarily converge.
The 1-base, $1, 2, 4, 8, 16, 32, 64...$ or $2^k, k \in \N$, we call the "root base" and provably converges. It is therefore a matter of determining how bases
transform into other bases that (hopefully) transform into the root base.

The following prime bases change bases to composite bases. So called "problem prime bases".
Up to first 1000 primes checked for this property. This removes the possibility that the prime bases always transform into other prime bases.
At this moment, the primes do not seem to harbor any special properties... which is a good thing.

``` 

```
23, 43, 73, 79, 83, 103, 107, 113, 139, 163, 173, 191, 193, 199, 223, 227, 233, 251, 263, 271, 283, 293, 313, 317, 331, 337, 353, 359, 367, 373, 383, 401, 419, 433, 443, 457, 463, 487, 491, 499, 503, 521, 523, 541, 557, 563, 569, 593, 599, 601, 613, 641, 643, 653, 659, 673, 683, 691, 709, 719, 733, 743, 751, 773, 797, 821, 823, 827, 863, 877, 883, 887, 919, 929, 937, 947, 953, 971, 983, 997, 1013, 1019, 1031, 1033, 1051, 1063, 1087, 1093, 1103, 1123, 1129, 1151, 1153, 1163, 1171, 1187, 1193, 1201, 1213, 1217, 1223, 1279, 1283, 1291, 1297, 1303, 1307, 1327, 1367, 1373, 1381, 1409, 1423, 1433, 1439, 1447, 1451, 1453, 1459, 1481, 1483, 1487, 1493, 1499, 1523, 1543, 1549, 1553, 1579, 1583, 1609, 1613, 1619, 1657, 1663, 1667, 1693, 1697, 1723, 1733, 1753, 1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1871, 1873, 1889, 1901, 1913, 1933, 1949, 1973, 1987, 1993, 1997, 2003, 2011, 2017, 2039, 2053, 2063, 2081, 2083, 2087, 2099, 2113, 2131, 2137, 2141, 2143, 2153, 2179, 2203, 2207, 2213, 2221, 2243, 2251, 2267, 2269, 2273, 2281, 2287, 2293, 2333, 2339, 2347, 2357, 2381, 2383, 2393, 2399, 2417, 2423, 2459, 2473, 2503, 2521, 2539, 2543, 2551, 2557, 2579, 2591, 2593, 2609, 2617, 2633, 2647, 2663, 2683, 2687, 2693, 2707, 2711, 2713, 2729, 2731, 2753, 2767, 2791, 2801, 2803, 2833, 2837, 2843, 2851, 2861, 2879, 2887, 2897, 2903, 2953, 2963, 2969, 2999, 3019, 3023, 3037, 3061, 3067, 3079, 3083, 3089, 3109, 3137, 3163, 3187, 3203, 3209, 3217, 3229, 3253, 3257, 3271, 3299, 3307, 3313, 3319, 3323, 3329, 3331, 3343, 3371, 3373, 3389, 3407, 3433, 3449, 3461, 3463, 3467, 3499, 3511, 3527, 3533, 3547, 3557, 3559, 3571, 3581, 3583, 3593, 3607, 3613, 3623, 3631, 3637, 3643, 3659, 3673, 3677, 3691, 3697, 3709, 3719, 3733, 3739, 3761, 3769, 3793, 3803, 3823, 3833, 3847, 3851, 3853, 3863, 3881, 3917, 3923, 3929, 3943, 3947, 3967, 3989, 4003, 4013, 4021, 4027, 4051, 4057, 4073, 4079, 4091, 4093, 4099, 4111, 4127, 4129, 4133, 4139, 4153, 4159, 4177, 4201, 4229, 4231, 4243, 4253, 4261, 4271, 4273, 4283, 4297, 4339, 4349, 4357, 4363, 4373, 4391, 4397, 4423, 4447, 4451, 4463, 4483, 4493, 4513, 4517, 4523, 4547, 4561, 4567, 4583, 4591, 4603, 4637, 4643, 4649, 4657, 4663, 4673, 4691, 4703, 4723, 4733, 4759, 4783, 4787, 4793, 4799, 4801, 4813, 4877, 4889, 4903, 4909, 4919, 4931, 4933, 4937, 4943, 4951, 4957, 4973, 4993, 5003, 5009, 5021, 5023, 5077, 5081, 5087, 5107, 5113, 5119, 5147, 5153, 5167, 5179, 5189, 5231, 5233, 5273, 5281, 5297, 5303, 5309, 5323, 5333, 5347, 5351, 5393, 5399, 5413, 5417, 5419, 5441, 5443, 5449, 5471, 5477, 5483, 5503, 5507, 5519, 5521, 5563, 5573, 5581, 5623, 5639, 5647, 5651, 5653, 5659, 5683, 5689, 5693, 5711, 5737, 5743, 5749, 5783, 5791, 5801, 5807, 5813, 5821, 5839, 5843, 5849, 5851, 5857, 5861, 5867, 5869, 5881, 5903, 5923, 5927, 5939, 5953, 5987, 6011, 6029, 6043, 6047, 6053, 6067, 6073, 6079, 6101, 6113, 6131, 6133, 6143, 6163, 6173, 6197, 6199, 6203, 6211, 6257, 6263, 6271, 6277, 6299, 6301, 6317, 6323, 6329, 6337, 6343, 6353, 6361, 6373, 6379, 6421, 6427, 6449, 6473, 6491, 6521, 6529, 6547, 6551, 6553, 6563, 6569, 6599, 6607, 6637, 6653, 6659, 6673, 6679, 6689, 6701, 6703, 6709, 6733, 6737, 6761, 6763, 6791, 6793, 6803, 6823, 6827, 6829, 6833, 6841, 6857, 6863, 6869, 6871, 6883, 6899, 6907, 6911, 6947, 6959, 6961, 6967, 6983, 7001, 7013, 7027, 7043, 7057, 7069, 7079, 7103, 7109, 7121, 7151, 7177, 7193, 7207, 7211, 7213, 7219, 7237, 7243, 7247, 7253, 7283, 7297, 7307, 7321, 7331, 7333, 7349, 7393, 7433, 7457, 7459, 7481, 7487, 7489, 7499, 7523, 7541, 7549, 7559, 7561, 7573, 7583, 7591, 7603, 7639, 7643, 7673, 7681, 7687, 7691, 7703, 7723, 7727, 7753, 7759, 7789, 7793, 7817, 7823, 7853, 7873, 7877, 7879, 7883, 7907, 7919
```